References
==========

.. [GenePlexus]
   * Liu, R., Mancuso, C.A. *et al.* (2020) `Supervised learning is an accurate method for network-based gene classification <https://doi.org/10.1093/bioinformatics/btaa150>`_. *Bioinformatics* 36, 3457–3465
   * Mancuso, C.A. *et. al.* (2024) `Joint representation of molecular networks from multiple species improves gene classification <https://doi.org/10.1371/journal.pcbi.1011773>`_. *PLOS Comp. Bio.* 20, e1011773
.. [GO]
   * Ashburner, M., Ball, C., Blake, J. *et al.* (2000) `Gene Ontology: tool for the unification of biology <https://doi.org/10.1038/75556>`_. *Nat Genet* 25, 25–29
   * The Gene Ontology Consortium (2021) `The Gene Ontology resource: enriching a GOld mine <https://doi.org/10.1093/nar/gkaa1113>`_. *Nucl. Acids Res.* 49, D325-D334
.. [Monarch]
   * Putnam, T.E. *et al.* (2024) `The Monarch Initiative in 2024: an analytic platform integrating phenotypes, genes and diseases across species <https://doi.org/10.1093/nar/gkad1082>`_. *Nucl. Acids Res.* 52, D938-D949
.. [Mondo]
   * , T.E. *et al.* (2024) `Mondo: Unifying diseases for the world, by the world <https://doi.org/10.1101/2022.04.13.22273750>`_. *medRxiv.*
.. [BioGRID]
   * Oughtred, R. *et al.* (2020) `The BioGRID database: A comprehensive biomedical resource of curated protein, genetic, and chemical interactions <https://doi.org/10.1002/pro.3978>`_. *Protein Sci.*
   * Stark, C. *et al.* (2006) `BioGRID: a general repository for interaction datasets <https://doi.org/10.1093/nar/gkj109>`_. *Nucl. Acids Res.* 34, D535–D539
.. [STRING]
   * Szklarczyk, D. *et al.* (2015) `STRING v10: protein–protein interaction networks, integrated over the tree of life <https://doi.org/10.1093/nar/gku1003>`_. *Nucl. Acids Res.* 43, D447–D452
.. [IMP]
   * Wong, A.K. *et al.* (2012) `IMP: a multi-species functional genomics portal for integration, visualization and prediction of protein functions and networks <https://doi.org/10.1093/nar/gks458>`_. *Nucl. Acids Res.* 40, W484–W490
.. [node2vec]
   * Grover, A., Leskovec, J. (2016) `Node2vec: Scalable Feature Learning for Networks <https://doi.org/10.1145/2939672.2939754>`_. *KDD '16* 855–864
.. [EggNOG]
   * Huerta-Cepas J. (2019) `eggNOG 5.0: a hierarchical, functionally and phylogenetically annotated orthology resource based on 5090 organisms and 2502 viruses <https://doi.org/10.1093/nar/gky1085>`_. *Nucl. Acids Res.* 47, D309–D314