References
==========

.. [GenePlexus]
   * Liu, R., Mancuso, C.A. *et al.* (2020) `Supervised learning is an accurate method for network-based gene classification <https://doi.org/10.1093/bioinformatics/btaa150>`_. *Bioinformatics* 36, 3457–3465
.. [GO]
   * Ashburner, M., Ball, C., Blake, J. *et al.* (2000) `Gene Ontology: tool for the unification of biology <https://doi.org/10.1038/75556>`_. *Nat Genet* 25, 25–29
   * The Gene Ontology Consortium (2021) `The Gene Ontology resource: enriching a GOld mine <https://doi.org/10.1093/nar/gkaa1113>`_. *Nucl. Acids Res.* 49, D325-D334
.. [DisGeNet]
   * Piñero, J. *et al.* (2019) `The DisGeNET knowledge platform for disease genomics: 2019 update <https://doi.org/10.1093/nar/gkz1021>`_. *Nucl. Acids Res.* 48, D845-D855
   * Piñero, J. *et al.* (2017) `DisGeNET: a comprehensive platform integrating information on human disease-associated genes and variants <https://doi.org/10.1093/nar/gkw943>`_. *Nucl. Acids Res.* 45, D833-D839
   *  Piñero, J. *et al.* (2015) `DisGeNET: a discovery platform for the dynamical exploration of human diseases and their genes <https://doi.org/10.1093/database/bav028>`_. *Database*
.. [BioGRID]
   * Oughtred, R. *et al.* (2020) `The BioGRID database: A comprehensive biomedical resource of curated protein, genetic, and chemical interactions <https://doi.org/10.1002/pro.3978>`_. *Protein Sci.*
   * Stark, C. *et al.* (2006) `BioGRID: a general repository for interaction datasets <https://doi.org/10.1093/nar/gkj109>`_. *Nucl. Acids Res.* 34, D535–D539
.. [STRING]
   * Szklarczyk, D. *et al.* (2015) `STRING v10: protein–protein interaction networks, integrated over the tree of life <https://doi.org/10.1093/nar/gku1003>`_. *Nucl. Acids Res.* 43, D447–D452
.. [STRING-EXP]
   * Szklarczyk, D. *et al.* (2015) `STRING v10: protein–protein interaction networks, integrated over the tree of life <https://doi.org/10.1093/nar/gku1003>`_. *Nucl. Acids Res.* 43, D447–D452
.. [GIANT-TN]
   * Greene, C., Krishnan, A., Wong, A. *et al.* (2015) `Understanding multicellular function and disease with human tissue-specific networks <https://doi.org/10.1038/ng.3259>`_. *Nat Genet* 47, 569–576
.. [node2vec]
   * Grover, A., Leskovec, J. (2016) `Node2vec: Scalable Feature Learning for Networks <https://doi.org/10.1145/2939672.2939754>`_. *KDD '16* 855–864
.. [Huang]
   * Huang, J.K. *et al.* (2018) `Systematic Evaluation of Molecular Networks for Discovery of Disease Genes <https://doi.org/10.1016/j.cels.2018.03.001>`_. *Cell Systems* 6 484–495
.. [MyGeneInfo]
   * Wu C, MacLeod I, Su AI* (2013) `BioGPS and MyGene.info: organizing online, gene-centric information <https://doi.org/10.1093/nar/gks1114>`_. *Nucl. Acids Res.* 43, D561-D565
   * Xin J, et al.* (2016) `High-performance web services for querying gene and variant annotation <https://doi.org/10.1186/s13059-016-0953-9>`_. Genome Biology, 17, 1-7
.. [DiseaseOntology]
   * Schriml, L.M. et al.* (2019) `Human Disease Ontology 2018 update: classification, content and workflow expansion <https://doi.org/10.1093/nar/gky1032>`_. *Nucl. Acids Res.* 47 D955-D962