{% extends "base.html" %}

{% block app_content %}
    <div class="container">
        <h1 id="geneplexus-help-page">GenePlexus <sup>&beta;eta</sup> Help Page</h1>
        <p>The help page provides various help, advice, and interpretation of the various inputs and outputs on the GenePlexus webserver. </p>
        <h2 id="inputs">Inputs</h2>
        <ul>
        <li><a href="#supplying-input-genes">Supplying input genes</a></li>
        <li><a href="#choosing-a-molecular-network">Choosing a molecular network</a></li>
        <li><a href="#choosing-a-way-to-represent-network-features">Choosing a way to represent network features</a></li>
        <li><a href="#choosing-negative-genes">Choosing negative genes</a></li>
        </ul>
        <h3 id="supplying-input-genes">Supplying input genes</h3>
        <p>The user can supply a list of genes by either uploading a file of genes or manually entering the genes in a text box. In both cases, each gene needs to be on a new line. 
        <p>Here is an example gene list file : <a href="{{url_for('static',filename='example-geneplexus-file.txt')}}" download>download example-geneplexus-file.txt</a> </p>    
        <p>The following types of gene IDs are allowed as inputs:</p>
        <ul>
        <li>Entrez</li>
        <li>Symbol</li>
        <li>ENSG (Ensembl gene)</li>
        <li>ENSP (Ensembl protein)</li>
        <li>ENST (Ensembl transcript)</li>
        </ul>
        <p>Note: The <a href="https://useast.ensembl.org/index.html">Ensembl</a> IDs (ENSG, ESNP, ENST) cannot contain any version numbers.</p>
        <p>The genes supplied by the user will be converted to <a href="https://www.ncbi.nlm.nih.gov/Web/Search/entrezfs.html">Entrez</a> gene identifiers using gene ID mappings obtained from 60,000 locally downloaded human gene files from <a href="https://mygene.info/">mygene.info</a>.</p>
        <p>After uploading/entering the genes, users can validate their input list by clicking the <code>Check if IDs are Valid</code> button. This will pop up two tables:</p>
        <ul>
        <li>A table that shows the original gene IDs the user supplied and their corresponding Entrez gene IDs. The last four columns indicate whether or not the gene is present in each of the networks in GenePlexus.</li>
        <li>A table that shows for each network, the total number of genes in that network and how many among them are also part of the user supplied gene list.</li>
        </ul>
        <h3 id="choosing-a-molecular-network">Choosing a molecular network</h3>
        <p>The user can choose between four human genome-scale molecular networks. For each network, the node are mapped into Entrez gene IDs. For the GIANT-TN network, the original network was additionally filtered to remove edges with scores below the prior probability (0.01). </p>
        <ul>
        <li><strong><a href="https://thebiogrid.org/">BioGRID</a></strong> (19,022 genes, 484,356 edges). This network is a relatively sparse unweighted network with high gene coverage. The edges in the network are only from verified experimental results.</li>
        <li><strong><a href="https://string-db.org/">STRING</a></strong> (18,582 genes, 5,521,113 edges). This network is a moderately well-connected weighted network that integrates information from multiple sources (experimental results, homology, annotation databases, etc).</li>
        <li><strong><a href="https://string-db.org/">STRING-EXP</a></strong> (17,417 genes, 2,121,428 edges). This network contains only the &quot;experimental&quot; edges from the STRING network.</li>
        <li><strong><a href="http://giant.princeton.edu/">GIANT-TN</a></strong> (25,689 genes, 38,904,929 edges). This is a nearly fully-connected weighted network that is derived from combining data from co-expression networks generated across 1000s of gene expression experiments along with gene interaction information from a few other sources including protein interactions. In this webserver, we use the tissue-naive GIANT network.</li>
        </ul>
        <h3 id="choosing-a-way-to-represent-network-features">Choosing a way to represent network features</h3>
        <p>When a network is to be used in a supervised machine learning model, the network connections can be represented as features in one of three ways:</p>
        <ul>
        <li><strong>Adjacency</strong>. This method uses the network edges directly from the molecular networks. This feature type might work best for the <code>STRING</code> and <code>STRING-EXP</code> networks.</li>
        <li><strong>Influence</strong>. The edge information in the original network is diffused over all edges using a random walk with restart technique. This feature type works best for dense networks where many edges may be a bit “noisy”, so here it might be best to use with <code>GIANT-TN</code>.</li>
        <li><strong>Embedding</strong>. The connectivity of a given gene in the network to all the other genes is represented with a low dimensional numerical vector using the <a href="https://arxiv.org/abs/1607.00653">node2vec</a> algorithm. This feature type works best with sparse networks, so here it might be best to use with <code>BioGRID</code>.</li>
        </ul>
        <h3 id="choosing-negative-genes">Choosing negative genes</h3>
        <p>In the supervised machine learning model, any gene from the user-supplied gene list that is able to be converted to an Entrez ID and is also in the network is considered part of the positive class. The user can then choose if they want to define genes in the negative class based on one of two geneset collections; <a href="http://geneontology.org/">Gene Ontology Biological Processes</a> or <a href="https://www.disgenet.org/">DisGeNet</a> based on whether the input genes represent a cellular process/pathway or a disease.</p>
        <p>GenePlexus then automatically selects the genes in the negative class by:</p>
        <ol>
        <li>Retaining all terms in the selected geneset collection that have between 10 and 200 genes annotated to them.</li>
        <li>Considering the total pool of possible negative genes to be any gene that has an annotation to at least one of the terms in the selected geneset collection.</li>
        <li>Removing genes that are in the positive class.</li>
        <li>Performing a hypergeometric test between the genes in the positive class and the lists of genes annotated to every term in the selected geneset collection. If the value of this hypergeometric test is less than 0.05, all genes from the given term are also removed from the pool of possible negative genes.</li>
        <li>Declaring all the remaining genes in the pool of possible negative genes as the negative class.</li>
        </ol>
        <p>Looking at the score for the cross-validation on the top of the results page may also help the user choose which network, feature type, and negative class is best for their gene list. Below is a figure showing a summary of results generated from our recent <a href="https://academic.oup.com/bioinformatics/article/36/11/3457/5780279">work</a> benchmarking this method.</p>
        <p><img src="{{url_for('static', filename='help_page_fig.png')}}" alt="help_page_fig" style="width:1100px;height:550px;" />
        <em>This figure displays results from the paper <a href="https://academic.oup.com/bioinformatics/article/36/11/3457/5780279?login=true">Supervised learning is an accurate method for network-based gene classification</a> where the parameters are options in the GenePlexus webserver. The left panel is for models trained for Gene Ontology biological prcoesses and the right panel is for modles trained for DisGeNet diseases. Each boxplot contains the results of anywhere between 89 and 160 models. Each model was trained using the study-bias holdout method, where for each geneset in a geneset collection, the most well studied genes were used to train the model and the least studied genes were used for testing. This figure can be used to help users pick the network, feature type and negative selection class that best suits their input gene list. It is worth noting that the STRING network is built in part by using Gene Ontology and DisGeNet annotations, and this circualriy could be the reason for the enhanced performance of the STRING network in this evaluation.</em>  </p>
        <h2 id="outputs">Outputs</h2>
        <ul>
        <li><a href="#information-at-the-top">Information at the top</a></li>
        <li><a href="#gene-predictions">Gene predictions</a></li>
        <li><a href="#comparison-to-other-pre-trained-models">Comparison to other pre-trained models</a></li>
        <li><a href="#network-graph">Network graph</a></li>
        <li><a href="#validation-data">Validation data</a></li>
        </ul>
        <h3 id="information-at-the-top">Information at the top</h3>
        <p>The box at the top displays the parameters of the job along with the job name. If the number of genes in the positive class is greater than 20, then the median value of the 5-fold cross validation performance is displayed in terms of <code>log2(auPRC/prior)</code>. This metric reports the 2-fold change of the area under the precision-recall curve (auPRC) over the prior auPRC expected by random chance. For example, a value of 1 indicates a model that performs twice as good as the expected result.</p>
        <h3 id="gene-predictions">Gene predictions</h3>
        <p>This table contains the prediction probability for every gene in the network, indicating its membership in the positive class. The probabilities are bounded from 0 to 1 with 1 being the highest probability of being in the positive class.</p>
        <p>The possible Known/Novel column values are:</p>
        <ul>
        <li>Known: These are genes that are in the positive class (supplied by the user).</li>
        <li>Novel: These are genes that are the predicted novel associations. These genes could have been part of the negative class or not used in training the model at all.</li>
        </ul>
        <p>The possible Class-labels column values are:</p>
        <ul>
        <li>P: Gene was part of the positive class during training.</li>
        <li>N: Gene was part of the negative class during training.</li>
        <li>U: Gene was unused during training.</li>
        </ul>
        <p>The table can be filtered by using the text box above the table. For example, often the top predictions are from genes in the positive class. So, it might be useful to type “Novel” into the text box to see what are the top predicted novel genes.</p>
        <p>Clicking the Entrez Gene ID takes the user to a webpage containing more information on that gene.</p>
        <h3 id="comparison-to-other-pre-trained-models">Comparison to other pre-trained models</h3>
        <p>The tables here show how similar the custom model trained on the user-supplied gene list compares to other pre-trained models. The webserver stores the model weights from machine-learning models trained using lists of genes annotated to numerous Gene Ontology biological processes and DisGeNet diseases. The similarity score is determined by:</p>
        <ol>
        <li>Finding the cosine similarity between the model trained with the user-supplied gene list and all the model weights from the genesets in a geneset collection, where the model was trained using the same molecular network and feature type selected by the user. </li>
        <li>Reading in a pre-saved matrix containing the similarities between models from genesets in the geneset collection that was chosen to make the negative class and models for genesets in the geneset collection being displayed in the table.</li>
        <li>Adding the vector of similarities found in step one as a row in this matrix.</li>
        <li>Finding both the columnwise and rowwise z-scores for each element in this appended row where the columnwise z-score corrects for the background among terms using that negative set of genes and the rowwise z-score corrects for the background among all genesets displayed in the table.</li>
        <li>Calcualting the final similarity as the combination of these two z-scores combined using the l2-norm.</li>
        </ol>
        <p>Clicking the Term ID takes the user to a page containing more information about that term.</p>
        <h3 id="network-graph">Network graph</h3>
        <p>The top predictions from the supervised machine learning model are displayed, showing how they are connected to each other in the network used to train the machine learning model.</p>
        <ul>
        <li>The nodes are colored by which class they were in, with green indicating genes in the positive class, blue indicating genes not used in training, and red indicating genes in the negative class.</li>
        <li>The size of the nodes scales with the prediction probability, with the larger nodes having the higher prediction probability.</li>
        </ul>
        <h3 id="validation-data">Validation data</h3>
        <p>The table here shows how the user input gene list was converted to Entrez IDs and if the converted Entrez IDs were in the network that was used to train the model.</p>
        <h2 id="other-information">Other Information</h2>
        <h3 id="references">References</h3>
        <h4 id="networks">Networks</h4>
        <h5 id="giant">GIANT</h5>
        <ul>
        <li>Greene CS, Krishnan A, Wong AK, Ricciotti E, Zelaya RA, Himmelstein DS, Zhang R, Hartmann BM, Zaslavsky E, Sealfon SC, Chasman DI, FitzGerald GA, Dolinski K, Grosser T, Troyanskaya OG. (2015) Understanding multicellular function and disease with human tissue-specific networks. <em>Nature Genetics</em> 47:569-576.<h5 id="string-and-string-exp">STRING and STRING-EXP</h5>
        </li>
        <li>Szklarczyk D, Franceschini A, Wyder S, Forslund K, Heller D, Huerta-Cepas J, Simonovic M, Roth A, Santos A, Tsafou KP, Kuhn M, Bork P, Jensen LJ, von Mering C. (2015) STRING v10: protein-protein interaction networks, integrated over the tree of life. <em>Nucleic Acids Research</em> 43:D447-452.<h5 id="biogrid">BioGRID</h5>
        </li>
        <li>Stark C, Breitkreutz BJ, Reguly T, Boucher L, Breitkreutz A, Tyers M. (2006) BioGRID: a general repository for interaction datasets. <em>Nucleic Acids Research</em> 34:D535-539.</li>
        </ul>
        <h4 id="geneset-collections">Geneset collections</h4>
        <h5 id="gene-ontology">Gene Ontology</h5>
        <ul>
        <li>Ashburner M, Ball CA, Blake JA, Botstein D, Butler H, Cherry JM, Davis AP, Dolinski K, Dwight SS, Eppig JT, Harris MA, Hill DP, Issel-Tarver L, Kasarskis A, Lewis S, Matese JC, Richardson JE, Ringwald M, Rubin GM, Sherlock G. (2000) Gene ontology: tool for the unification of biology. The Gene Ontology Consortium. <em>Nature Genetics</em> 25:25-29.</li>
        <li>The Gene Ontology Consortium. (2019) The Gene Ontology Resource: 20 years and still GOing strong. <em>Nucleic Acids Research</em> 47:D330-338.<h5 id="disgenet">DisGeNet</h5>
        </li>
        <li>Piñero J, Queralt-Rosinach N, Bravo À, Deu-Pons J, Bauer-Mehren A, Baron M, Sanz F, Furlong LI. (2015) DisGeNET: a discovery platform for the dynamical exploration of human diseases and their genes. <em>Database</em> bav028.</li>
        <li>Piñero J, Bravo À, Queralt-Rosinach N, Gutiérrez-Sacristán A, Deu-Pons J, Centeno E, García-García J, Sanz F, Furlong LI. (2017) DisGeNET: a comprehensive platform integrating information on human disease-associated genes and variants. <em>Nucleic Acids Research</em> 45:D833-839.</li>
        <li>Schriml LM, Mitraka E, Munro J, Tauber B, Schor M, Nickle L, Felix V, Jeng L, Bearer C, Lichenstein R, Bisordi K, Campion N, Hyman B, Kurland D, Oates CP, Kibbey S, Sreekumar P, Le C, Giglio M, Greene C. (2019) Human Disease Ontology 2018 update: classification, content and workflow expansion. <em>Nucleic Acids Research</em> 47:D955-D962.<h4 id="mygene-info">Mygene.info</h4>
        </li>
        <li>Wu C, MacLeod I, Su AI (2013) BioGPS and MyGene.info: organizing online, gene-centric information. <em>Nucleic Acids Research</em> 41:D561-D565. </li>
        <li>Xin J, Mark A, Afrasiabi C, Tsueng G, Juchler M, Gopal N, Stupp GS, Putman TE, Ainscough BJ, Griffith OL, Torkamani A, Whetzel PL, Mungall CJ, Mooney SD, Su AI, Wu C (2016) High-performance web services for querying gene and variant annotation. <em>Genome Biology</em> 17:1-7.</li>
        </ul>
        <h3 id="license">License</h3>
        <p>The results of the GenePlexus webserver are licensed under the <a href="https://creativecommons.org/licenses/by/4.0/legalcode">Creative Commons License: Attribution 4.0 International</a></p>
        <h4 id="license-information-for-networks-and-genesets-collections-used-in-this-work">License information for networks and genesets collections used in this work</h4>
        <ul>
        <li>STRING, STRING-EXP and GO use the <a href="https://creativecommons.org/licenses/by/4.0/legalcode">Creative Commons License: Attribution 4.0 International</a>. </li>
        <li>DisGenet uses the <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/legalcode">Creative Commons License: Attribution-NonCommercial-ShareAlike 4.0 International</a>.</li>
        <li>BioGRID and MyGene.info have terms of use on their websites (see below).</li>
        </ul>
        <p>The location of the license agreement or terms-of-use for each network or geneset collection can be found by clicking the following links. We note that no license agreement was available for GIANT, however we have obtained permission from the owner of the material to redistribute the network and they have noted that they will be adding a <a href="https://creativecommons.org/licenses/by-nc/4.0/legalcode">Creative Commons License: Attribution-NonCommercial 4.0 International</a> to the website soon.</p>
        <ul>
        <li><a href="https://string-db.org/cgi/access.pl?footer_active_subpage=licensing">STRING and STRING-EXP</a></li>
        <li><a href="https://wiki.thebiogrid.org/doku.php/terms_and_conditions">BioGRID</a></li>
        <li><a href="http://geneontology.org/docs/go-citation-policy/">GO</a></li>
        <li><a href="http://www.disgenet.org/legal">DiGeNet</a></li>
        <li><a href="https://mygene.info/terms">MyGene.info</a></li>
        </ul>
    </div>
{% endblock %}